[0017] rfc: transaction valid since (#83)
nervosnetwork/rfcs: rfcs/0017-tx-valid-since/0017-tx-valid-since.md

---
Number: "0017"
Category: Standards Track
Status: Proposal
Author: Jinyang Jiang
Organization: Nervos Foundation
Created: 2019-03-11
---

# Transaction valid since

## Abstract

This RFC suggests adding a new consensus rule to prevent a cell to be spent before a certain block timestamp or a block number.

## Summary 

> Transaction adds a new `u64` type field `valid_since`, which prevent the transaction to be mined before a certain point, the highest bit of `valid_since` is `type flag`, the other 63 bits represent the actual value `V`, the consensus to validate this field described as follow:

Reviewer: ```suggestion
Transaction adds a new `u64` type field `valid_since`, which prevents the transaction to be mined before a certain point, the highest bit of `valid_since` is `type flag`, the other 63 bits represent the actual value `V`, the consensus to validate this field described as follow:
```

> * ignore this validate rule if all bits of `valid_since` is 0.

Reviewer: ```suggestion
* ignore this validate rule if all bits of `valid_since` are 0.
```

> * `V` represent a block number when `type flag` is 0, the validation MUST failed if `tip.block_number < V`.

Reviewer: ```suggestion
* `V` represents a block number when `type flag` is 0, the validation MUST fail if `tip.block_number < V`.
```

> * `V` represent a block timestamp when `type flag` is 1, the timestamp represented as `V * 512` seconds, the validation MUST failed if `tip.timestamp < V * 512`.

Reviewer: ```suggestion
* `V` represents a block timestamp when `type flag` is 1, the timestamp is represented as `V * 512` seconds, the validation MUST fail if `tip.timestamp < V * 512`.
```

> Otherwise, the transaction validate SHOULD continue.

Reviewer: ```suggestion
Otherwise, the transaction validation SHOULD continue.
```
